lib/virsh_autotest_base.pm

```diff
@@ -0,0 +1,126 @@
+# VIRSH TEST MODULE BASE PACKAGE
+#
+# Copyright © 2019 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved. This file is offered as-is,
+# without any warranty.
+#
+# Summary: This is the base package for virsh test modules, for example,
+# tests/virtualization/xen/hotplugging.pm
+# tests/virt_autotest/virsh_internal_snapshot.pm
+# tests/virt_autotest/virsh_external_snapshot.pm
+# and etc.
+# Maintainer: Wayne Chen <wchen@suse.com>
+
+package virsh_autotest_base;
+
+use base "consoletest";
+use strict;
+use warnings;
+use POSIX 'strftime';
+use File::Basename;
+use Data::Dumper;
+use XML::Writer;
+use IO::File;
+use testapi;
+use utils;
+use virt_utils;
+use xen;
+
+sub run_test {
+    die('Please override this subroutine in children modules to run desired tests.');
+}
+
+sub run {
+    my ($self) = @_;
+    $self->{"start_run"} = time();
+    $self->run_test;
+    $self->{"stop_run"} = time();
+    #(caller(0))[3] helps pass calling subroutine name into called subroutine
+    $self->junit_log_provision((caller(0))[3]) if get_var("VIRT_AUTOTEST");
+}
+
+sub junit_log_provision {
+    my ($self, $runsub) = @_;
+    my $status    = eval { $runsub =~ /post_fail_hook/img ? 'FAILED' : 'PASSED' };
+    my $tc_result = $self->analyzeResult($status);
+    $self->junit_log_params_provision($tc_result);
+    ###Load instance attributes into %stats
+    my %stats;
+    foreach (keys %{$self}) {
+        if (defined($self->{$_})) {
+            if (ref($self->{$_}) eq 'HASH') {
+                %{$stats{$_}} = %{$self->{$_}};
+            }
+            elsif (ref($self->{$_}) eq 'ARRAY') {
+                @{$stats{$_}} = @{$self->{$_}};
+            }
+            else {
+                $stats{$_} = $self->{$_};
+            }
+        }
+        else {
+            next;
+        }
+    }
```

Comment:
What about puttting line 50-67 also into function junit_log_params_provision? Or a separate function if you think not suitable.

Reply:
Not very convincing, IMHO.

Reply:
Well, please give the reason here.

Reply:
foreach loop is already a very compact form to iterate the things you want. So you not need to create a new subroutine to wrap up only "foreach" in it, although foreach might be called more than once. It does not only involves a new subroutine, it is also related to class instance. So, again, "Any modification here requires further tuning or tweak in other subroutines or modules. This is beyond the scope of this PR. In fact, there are many imperfection in the original implementation, so it should be kept as what it was currently in this PR. This is what I mean."

Reply:
I agree with @alice-suse.

Reply:
Actually, it does not deserve a new subroutine in lib file IMHO. Because according to subroutine definition:
##################################################################
A subroutine (or routine, function, procedure, macro, etc.) is, at its heart, a named chunk of work. It’s shorthand that allows you to think about your problem in bigger chunks. Bigger chunks means two things:

    You can break the problem up into smaller problems that you can solve independently.
    You can use these solutions to solve your overall problem with greater confidence.
#####################################################################
So you can see the foreach loop itself is already an "atomic problem" itself. If it is wrapped up in another subroutine, it does not improve or alter anything. It is just moved to another place and given a new name. And at the same time, it decreases the "confidence", because it does not reflect the operations that are closely related to class instance anymore. It stands in between, half like a common function, half like a class member, and it might not be called by other modules at all. It becomes useless, so it decreases "confidence".

It should be improved together with upcoming feature implementations, because it is closely related to class instance. According to your saying, yes, "nothing is perfect" and "we try our best", these two do not conflict with each other, and sometimes we need to accept it before evolving to the perfection.

Additionally these 10 lines might be useless to all other modules or libs. It should be resolved further within class implementation IMHO. I can classify all these to "junit log enhancement" from what I have seen till now. So any upcoming improvements fall within this category deserve a new or separate PR, which also makes more sense.
@pdostal